consensus/src/dag/mod.rs

```diff
@@ -0,0 +1,641 @@
+//! Converts units from the network into ones that are in the Dag, in the correct order.
+use std::collections::HashMap;
+
+use crate::{
+    alerts::{Alert, ForkingNotification},
+    units::{
+        SignedUnit, UncheckedSignedUnit, Unit, UnitStore, Validator as UnitValidator, WrappedUnit,
+    },
+    Data, Hasher, MultiKeychain,
+};
+use log::{debug, trace, warn};
+
+mod reconstruction;
+mod validation;
+
+pub use reconstruction::{ReconstructedUnit, Request};
+use reconstruction::{Reconstruction, ReconstructionResult};
+pub use validation::ValidatorStatus as DagStatus;
+use validation::{Error as ValidationError, Validator};
+
+const LOG_TARGET: &str = "AlephBFT-dag";
+
+/// The result of sending some information to the Dag.
+pub struct DagResult<H: Hasher, D: Data, MK: MultiKeychain> {
+    units: Vec<ReconstructedUnit<SignedUnit<H, D, MK>>>,
+    requests: Vec<Request<H>>,
+    alerts: Vec<Alert<H, D, MK::Signature>>,
+}
+
+impl<H: Hasher, D: Data, MK: MultiKeychain> DagResult<H, D, MK> {
+    fn empty() -> Self {
+        DagResult {
+            units: Vec::new(),
+            requests: Vec::new(),
+            alerts: Vec::new(),
+        }
+    }
+
+    fn alert(alert: Alert<H, D, MK::Signature>) -> Self {
+        DagResult {
+            units: Vec::new(),
+            requests: Vec::new(),
+            alerts: vec![alert],
+        }
+    }
+
+    fn add_alert(&mut self, alert: Alert<H, D, MK::Signature>) {
+        self.alerts.push(alert);
+    }
+
+    fn accumulate(&mut self, other: DagResult<H, D, MK>) {
+        let DagResult {
```

Comment:
Could've used `(units, requests, alerts) = other.into()` (assumming you will keep `impl From`s)